contracts/src/v0.8/dev/VRFSubscriptionBalanceMonitor.sol

```diff
@@ -0,0 +1,314 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.6;
+
+import "../ConfirmedOwner.sol";
+import "../interfaces/KeeperCompatibleInterface.sol";
+import "../interfaces/VRFCoordinatorV2Interface.sol";
+import "../interfaces/LinkTokenInterface.sol";
+import "@openzeppelin/contracts/security/Pausable.sol";
+
+/**
+ * @title The VRFSubscriptionBalanceMonitor contract.
+ * @notice A keeper-compatible contract that monitors and funds VRF subscriptions.
+ */
+contract VRFSubscriptionBalanceMonitor is ConfirmedOwner, Pausable, KeeperCompatibleInterface {
+  VRFCoordinatorV2Interface public COORDINATOR;
+  LinkTokenInterface public LINKTOKEN;
+
+  uint256 private constant MIN_GAS_FOR_TRANSFER = 55_000;
+
+  event FundsAdded(uint256 amountAdded, uint256 newBalance, address sender);
+  event FundsWithdrawn(uint256 amountWithdrawn, address payee);
+  event TopUpSucceeded(uint64 indexed subscriptionId);
+  event TopUpFailed(uint64 indexed subscriptionId);
+  event KeeperRegistryAddressUpdated(address oldAddress, address newAddress);
+  event VRFCoordinatorV2AddressUpdated(address oldAddress, address newAddress);
+  event LinkTokenAddressUpdated(address oldAddress, address newAddress);
+  event MinWaitPeriodUpdated(uint256 oldMinWaitPeriod, uint256 newMinWaitPeriod);
+
+  error InvalidWatchList();
+  error OnlyKeeperRegistry();
+  error DuplicateSubcriptionId(uint64 duplicate);
+
+  struct Target {
+    bool isActive;
+    uint96 minBalanceJuels;
+    uint96 topUpAmountJuels;
+    uint56 lastTopUpTimestamp;
+  }
+
+  address private s_keeperRegistryAddress;
+  uint256 private s_minWaitPeriodSeconds;
+  uint64[] private s_watchList;
+  mapping(uint64 => Target) internal s_targets;
```

Comment:
These could use some doc comments to aid in understanding.

Reply:
I wonder if you need a `mapping` and a `uint64[]` to do the operations. Could you do with just a `Target[]`? Everytime `setWatchList` is called the existing data is erased and overwritten. Might also make `topUp` cheaper but not sure (unless we can't do `Target[] memory targets = s_targets` in a single `SLOAD`)?